### PR TITLE
WIP Dynamic pipelines

### PR DESCRIPTION
**Pull request by podshumok**

### CODE REVIEW

File: slep002/proposal.rst [hunk 1]
```diff
@@ -0,0 +1,559 @@
...
+
+.. contents:: Table of contents
+   :depth: 3
+
+Goals
+=====
+
+* Being backward-compatible
+* Allow interactive pipeline construction (for example in IPython)
+* Support adding and replacing parts of pipeline
```
Comment (amueller) [inline slep002/proposal.rst]: We actually recently merged a PR that allows replacing parts of a pipeline

File: slep002/proposal.rst [hunk 2]
```diff
@@ -0,0 +1,559 @@
...
+    pipe.steps.append(['name', estimator])
+
+The only difference is that special handler is returned instead of `None`.
+
+Enhanced: by indexing
+.....................
+
+Using dict-like syntax if very user-friendly::
+
+    pipe.steps['name'] = estimator
```
Comment (amueller) [inline slep002/proposal.rst]: this is now possible

File: slep002/proposal.rst [hunk 3]
```diff
@@ -0,0 +1,559 @@
...
+
+Enhanced: `add` function
+........................
+
+Alias to previous two calls::
+
+    pipe.steps.add('name', estimator)
```
Comment (amueller) [inline slep002/proposal.rst]: to the end? Whats the benefit of that?

**Comment by amueller** [comment]
btw, maybe you're interested in this gist: https://gist.github.com/amueller/643f812a275a9e0c75048aab6988a92c

File: slep002/proposal.rst (hunk 3 above)
Comment (podshumok) [inline slep002/proposal.rst]: just to call append_func(name, est) instead of append_func((name, est))

Comment (amueller) [inline slep002/proposal.rst]: I meant in what situation do you want to add an estimator to the end of a pipeline. That seems like a weird operation to me. Can you give a usage example?

File: slep002/proposal.rst (hunk 1 above)
Comment (podshumok) [inline slep002/proposal.rst]: We'll probably will have to be compatible with that too :)

File: slep002/proposal.rst (hunk 3 above)
Comment (podshumok) [inline slep002/proposal.rst]: First, to dynamically construct the pipeline.
Second, you remove last estimator, get a transformer pipeline, transform train data, fit new (different) estimator with them, add fitted estimator to the end of the pipeline, getting fitted pipeline

**Comment by podshumok** [comment]
Personally, I don't see anything good in binary ops for estimators. And duck-typing will not work.

File: slep002/proposal.rst (hunk 2 above)
Comment (podshumok) [inline slep002/proposal.rst]: I'll update the title "Enhanced:" to ''Backward-compatible" after stable release.

File: slep002/proposal.rst (hunk 3 above)
Comment (amueller) [inline slep002/proposal.rst]: I don't understand the first. What do you mean by that? For the second: that's the same as changing the step, right?

**Comment by amueller** [comment]
> Personally, I don't see anything good in binary ops for estimators. And duck-typing will not work.

Can you elaborate both points? The gist actually implements a way to incrementally built a pipeline, which I think is what you're trying to do here.

**Comment by podshumok** [comment]
The problem is user-provided estimator doesn't have to be a subclass of BaseEstimator.

Even if somehow being a BaseEstomator will become a requirement for Pipeline steps explicit API for this should exist too

File: slep002/proposal.rst (hunk 3 above)
Comment (podshumok) [inline slep002/proposal.rst]: First:
```
>>> pipe = Pipeline()
>>> pipe.steps.add(...)
>>> pipe.steps.add(...)
# play with pipe with that two steps, explore params, do gridsearch
>>> pipe.steps.add(...)
# play with new pipe
>>> pipe.steps.replace('some', ...)
>>> pipe.steps.insert(0, imputer)
```
and so on

Second:
You save some fit steps, but generally - yes.

Another option is to add a classifier after a clusterer.

File: slep002/proposal.rst (hunk 2 above)
Comment (jnothman) [inline slep002/proposal.rst]: This is not possible: steps is still a list. Use `set_params`.

Comment (podshumok) [inline slep002/proposal.rst]: I'm proposing to introduce special internal data type/class/structure to support this

**Comment by amueller** [comment]
Can you give an example of a use-case that this solves that can't be solved right now?
For example, when do you need the label processing within a pipeline?

**Comment by podshumok** [comment]
One may want to do label encoding in the beginning of the pipeline, anywhere in the pipeline it may be necessary to apply some function to labels (to better suit following steps), clustering and classification can be chained into a pipeline (although at prediction stage first step, clustering, should be ignored)

Also, some examples are provided in the end of the document.

**Pull request merged**